jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+
+"""Wallet service
+
+The purpose of this independent service is to allow
+running applications to keep an up to date, asynchronous
+view of the current state of its wallet, deferring any
+polling mechanisms needed against the backend blockchain
+interface here.
+"""
+
+jlog = get_log()
+
+class WalletService(Service):
+
+    def __init__(self, wallet):
+        # The two principal member variables
+        # are the blockchaininterface instance,
+        # which is currently global in JM but
+        # could be more flexible in future, and
+        # the JM wallet object.
+        # TODO support multiple wallets here.
+        self.bci = jm_single().bc_interface
+        self.wallet = wallet
+        self.synced = False
+
+        # Dicts of registered callbacks, by type
+        # and then by txinfo, for events
+        # on transactions.
+        self.callbacks = {}
+        self.callbacks["all"] = []
+        self.callbacks["unconfirmed"] = {}
+        self.callbacks["confirmed"] = {}
+
+        self.restart_callback = None
+
+        # transactions we are actively monitoring,
+        # i.e. they are not new but we want to track:
+        self.active_txids = []
+
+    def startService(self):
+        """ Encapsulates start up actions.
+        Here wallet sync.
+        """
+        Service.startService(self)
+        self.request_sync_wallet()
+
+    def stopService(self):
+        """ Encapsulates shut down actions.
+        Here shut down main tx monitoring loop.
+        """
+        self.monitor_loop.stop()
+        Service.stopService(self)
+
+    def isRunning(self):
+        if self.running == 1:
+            return True
+        return False
+
+    def add_restart_callback(self, callback):
+        """ Sets the function that will be
+        called in the event that the wallet
+        sync completes with a restart warning.
+        The only argument is a message string,
+        which the calling function displays to
+        the user before quitting gracefully.
+        """
+        self.restart_callback = callback
+
+    def request_sync_wallet(self):
+        """ Ensures wallet sync is complete
+        before the main event loop starts.
+        """
+        d = task.deferLater(reactor, 0.0, self.sync_wallet)
+        d.addCallback(self.start_wallet_monitoring)
+
+    def sync_wallet(self, fast=True):
+        """ Syncs wallet using fast sync.
+        TODO: enable slow sync as option here.
+        Before starting the event loop, we cache
+        the current most recent transactions as
+        reported by the blockchain interface, since
+        we are interested in deltas.
+        """
+        self.bci.sync_wallet(self.wallet, fast=fast,
+                             restart_cb=self.restart_callback)
+        self.synced = self.bci.wallet_synced
+        # Don't attempt updates on transactions that existed
+        # before startup
+        self.old_txs = self.bci.list_transactions(100)
```

Comment:
Here `old_txs` is a list of transaction objects as returned from the rpc, but in `transaction_monitor` it is only a list of `txid`. Additionally it's not defined in `__init__`.

Reply:
Bug! That's what caused the erroneous "new tx processing" on startup. Thanks. Fixed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/e1fbc427120e31fdc3e96fe761720abf085390f2

Reply:
`old_txs` is still missing from `WalletService.__init__`